File: src/lib/api/scribe.ts
import { API_BASE_URL } from "./client";

export type ScribeSourceMode = "manual_transcript" | "audio";

export type PromptAnswer = {
  selected_value?: string;
  label?: string;
  text?: string;
};

export type PhysicianPromptResponseInput = {
  requestId: string;
  promptId: string;
  promptType: string;
  question: string;
  answer: PromptAnswer;
  encounterId: string;
  conversationHistory?: unknown[];
};

export type ProcessScribeInput = {
  patientRecordNumber: string;
  encounterId: string;
  consultationTime: string;
  sourceMode: ScribeSourceMode;
  manualTranscript?: string;
  audioFile?: File | null;
  doctorId?: string;
  languageHint?: string;
  dialectHint?: string;
  noteFormat?: string;
  privacyMode?: string;
  patientContextJson?: string;
};

export type LogprobToken = {
  token: string;
  logprob: number;
};

export type LogprobData = {
  avg_logprob: number | null;
  normalised_score: number | null;
  below_threshold: boolean;
  token_count: number;
  top_uncertain_tokens: LogprobToken[];
};

export type ScribeResponse = {
  request_id?: string;
  status?: string;
  pipeline_version?: string;
  audio?: unknown;
  speaker_context?: unknown;
  models_used?: unknown;
  inference?: unknown;
  transcription?: unknown;
  translation?: unknown;
  soap_note?: unknown;
  claims?: unknown;
  uncertainty?: unknown;
  uncertain_words?: unknown;
  uncertainty_spans?: unknown;
  physician_questions?: unknown;
  physician_prompts?: unknown;
  note_actions?: unknown;
  prompt_followup?: unknown;
  privacy?: unknown;
  frontend_hints?: unknown;
  providers_used?: unknown;
  pipeline?: Record<string, unknown>;
  // Pipeline B fields
  logprob_data?: LogprobData;
  self_eval?: unknown;
  question_eval?: unknown;
  mandatory_review?: boolean;
  partial_check?: boolean;
  warnings?: string[];
};

export type PromptResponse = {
  assistant_message?: string;
  note_update_suggestion?: unknown;
  claim_update_suggestion?: unknown;
  next_prompts?: unknown;
  models_used?: unknown;
  inference?: unknown;
  warnings?: unknown;
};

function appendIfPresent(formData: FormData, key: string, value?: string) {
  if (value && value.trim().length > 0) {
    formData.append(key, value);
  }
}

export async function processScribe(input: ProcessScribeInput): Promise<ScribeResponse> {
  const formData = new FormData();
  formData.append("patient_record_number", input.patientRecordNumber);
  formData.append("encounter_id", input.encounterId);
  formData.append("consultation_time", input.consultationTime);
  formData.append("source_mode", input.sourceMode);

  appendIfPresent(formData, "doctor_id", input.doctorId);
  appendIfPresent(formData, "language_hint", input.languageHint);
  appendIfPresent(formData, "dialect_hint", input.dialectHint);
  appendIfPresent(formData, "note_format", input.noteFormat);
  appendIfPresent(formData, "privacy_mode", input.privacyMode);
  appendIfPresent(formData, "patient_context_json", input.patientContextJson);

  if (input.sourceMode === "audio" && input.audioFile) {
    formData.append("audio_file", input.audioFile, input.audioFile.name || "consultation.webm");
  }

  if (input.sourceMode === "manual_transcript") {
    formData.append("manual_transcript", input.manualTranscript ?? "");
  }

  const response = await fetch(`${API_BASE_URL}/api/v1/scribe/process`, {
    method: "POST",
    body: formData
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(message || `SAJIL API request failed with ${response.status}`);
  }

  return response.json() as Promise<ScribeResponse>;
}

export async function respondToPhysicianPrompt(input: PhysicianPromptResponseInput): Promise<PromptResponse> {
  const response = await fetch(`${API_BASE_URL}/api/v1/scribe/prompt-response`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      request_id: input.requestId,
      prompt_id: input.promptId,
      prompt_type: input.promptType,
      question: input.question,
      answer: input.answer,
      scribe_context: {
        encounter_id: input.encounterId
      },
      conversation_history: input.conversationHistory ?? []
    })
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(message || `Prompt response failed with ${response.status}`);
  }

  return response.json() as Promise<PromptResponse>;
}

export async function fetchDemo(key: string): Promise<ScribeResponse> {
  const response = await fetch(`${API_BASE_URL}/api/v1/demo/${encodeURIComponent(key)}`);

  if (!response.ok) {
    const message = await response.text();
    throw new Error(message || `Demo fetch failed with ${response.status}`);
  }

  return response.json() as Promise<ScribeResponse>;
}

export async function runNoteAction(actionId: string) {
  const response = await fetch(`${API_BASE_URL}/api/v1/scribe/note-actions/${encodeURIComponent(actionId)}`, {
    method: "POST"
  });

  if (!response.ok) {
    const message = await response.text();
    throw new Error(message || `Note action failed with ${response.status}`);
  }

  return response.json() as Promise<{ status?: string }>;
}

File: api/app/routers/scribe.py
import os
from typing import Any, Dict, Optional

from fastapi import APIRouter, Body, UploadFile, File, Form, Header, HTTPException
from app.core.config import get_settings
from app.schemas.scribe import PromptAnswerRequest
from app.services.scribe_service import ScribeService

router = APIRouter()


def _build_scribe_service(
    x_debug_fanar_api_key: Optional[str] = None,
    x_debug_groq_api_key: Optional[str] = None,
) -> ScribeService:
    if x_debug_fanar_api_key:
        os.environ["FANAR_API_KEY"] = x_debug_fanar_api_key
    if x_debug_groq_api_key:
        os.environ["GROQ_API_KEY"] = x_debug_groq_api_key
    if x_debug_fanar_api_key or x_debug_groq_api_key:
        get_settings.cache_clear()
    return ScribeService()

@router.post("/process")
async def process_scribe(
    audio_file: Optional[UploadFile] = File(None),
    patient_record_number: str = Form(...),
    encounter_id: str = Form(...),
    consultation_time: str = Form(...),
    doctor_id: Optional[str] = Form(None),
    language_hint: str = Form("ar"),
    dialect_hint: str = Form("gulf"),
    note_format: str = Form("SOAP"),
    privacy_mode: str = Form("process_audio_delete_after"),
    source_mode: str = Form("audio"),
    patient_context_json: Optional[str] = Form(None),
    manual_transcript: Optional[str] = Form(None),
    x_debug_fanar_api_key: Optional[str] = Header(None),
    x_debug_groq_api_key: Optional[str] = Header(None),
):
    try:
        scribe_service = _build_scribe_service(x_debug_fanar_api_key, x_debug_groq_api_key)
        return await scribe_service.process(
            audio_file=audio_file,
            patient_record_number=patient_record_number,
            encounter_id=encounter_id,
            consultation_time=consultation_time,
            doctor_id=doctor_id,
            language_hint=language_hint,
            dialect_hint=dialect_hint,
            note_format=note_format,
            privacy_mode=privacy_mode,
            source_mode=source_mode,
            patient_context_json=patient_context_json,
            manual_transcript=manual_transcript,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/prompt-response")
async def answer_prompt(request: PromptAnswerRequest):
    try:
        scribe_service = _build_scribe_service()
        return scribe_service.respond_to_prompt(request.model_dump())
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/note-actions/{action_id}")
async def run_note_action(action_id: str, payload: Optional[Dict[str, Any]] = Body(None)):
    return {
        "action_id": action_id,
        "status": "under_development",
        "message": "This note-processing action is under development.",
        "payload_received": payload or {},
    }
